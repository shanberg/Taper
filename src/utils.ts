export function formatDate(date, style = 'conciseUS') {
let options;

  switch (style) {
    case 'conciseUS':
      options = { month: 'numeric', day: 'numeric' };
      break;
    case 'conciseEU':
      options = { day: 'numeric', month: 'numeric', year: '2-digit' };
      break;
    case 'longform':
      options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
      break;
    // Add more styles as needed
    default:
      options = {};
  }

  return new Intl.DateTimeFormat('default', options).format(new Date(date));
}

export const isRowValid = (row): boolean => {
  return row.dose <= 0 || row.daysForDose <= 0
}

export const isRowPlaceholder = (row): boolean => {
  return row.dose === 0 && row.daysForDose === 0
}


// Helper function to format dates as 'YYYY-MM-DD'
export function yyyymmdd(date) {
  const d = new Date(date),
        month = '' + (d.getMonth() + 1),
        day = '' + d.getDate(),
        year = d.getFullYear();

  return [year, month.padStart(2, '0'), day.padStart(2, '0')].join('-');
}