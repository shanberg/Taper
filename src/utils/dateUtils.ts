import { TaperDate } from '../TaperDate';

export const isDateThisYear = (date: ScheduleDate): boolean => {
  return date.getFullYear() === new Date().getFullYear();
};

/** Higher-order function to create a cached formatter */
function createCachedFormatter() {
  const dateTimeFormatCache = new Map();

  return function (date: ScheduleDate, lang: string = 'en-US'): LocaleDate {
    const yearFormat = isDateThisYear(date) ? undefined : 'numeric';
    const cacheKey = `${lang}-${yearFormat}`;

    // Check if the formatter is already in the cache
    if (!dateTimeFormatCache.has(cacheKey)) {
      // Create a new formatter and store it in the cache
      const formatter = new Intl.DateTimeFormat(lang, {
        month: 'short',
        day: 'numeric',
        year: yearFormat
      });
      dateTimeFormatCache.set(cacheKey, formatter);
    }

    // Use the cached formatter
    const formatter = dateTimeFormatCache.get(cacheKey);
    return formatter.format(date) as LocaleDate;
  };
}

// Create a cached version of formatDate
export const cachedFormatDate = createCachedFormatter();


export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}
