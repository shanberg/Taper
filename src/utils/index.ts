export * from './dateUtils';
export * from './stepUtils';
export * from './scheduleUtils';
export * from './serializationUtils';
export * from './textUtils';
export * from './languageUtils';