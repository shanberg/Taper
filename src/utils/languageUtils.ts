import { LANGUAGES } from '../consts';

export function getLanguageFromKey(languageKey: string): Language {
  return LANGUAGES.find((l) => l.lang === languageKey) || LANGUAGES[0];
}