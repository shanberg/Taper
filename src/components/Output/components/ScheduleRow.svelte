<script lang="ts">
	import { appStore } from '../../../stores';
	import FormStep from '../../FormStep.svelte';
	import AddStepButton from '../../AddStepButton.svelte';
	import { TaperDate } from '../../../TaperDate';
	import { getPeriodsWithDosesForStep, stepIsOrAfterPlaceholder, isStepInvalid, formatStepText, calculateScheduleSummary } from '../../../utils';
	import List from '../../List.svelte';
	import ListItem from '../../ListItem.svelte';

	export let schedule: Schedule;
  export let index: number;
  export let selectedLanguage: Language;
	
	let stepStartDate: ScheduleDate;
	let stepEndDate: ScheduleDate;
	
	$: displayMode = schedule.displayMode;
	$: scheduleStartDate = schedule.startDate;
	$: steps = schedule.steps;
	$: step = schedule.steps[index];
	$: isStepPlaceholder = step.dose === 0 && step.daysForDose === 0;
	$: isInvalid = !isStepPlaceholder && isStepInvalid(step);
	$: isLastPlaceholderStep = index === steps.length - 1 && isStepPlaceholder;

	// Calculate the start and end dates
	$: {
		const taperStartDate = new TaperDate(scheduleStartDate);
		const totalDaysForStartDate =
			steps
				.slice(0, index)
				.reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
			index;
		taperStartDate.incrementByDays(totalDaysForStartDate);
		const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
		taperEndDate.incrementByDays(step.daysForDose - 1);

		stepStartDate = taperStartDate.toScheduleDate();
		stepEndDate = taperEndDate.toScheduleDate();
	}

</script>

<div class="row" class:isInvalid class:isStepPlaceholder class:isLastPlaceholderStep>
	<span class="written-plan">
		{formatStepText({ step, stepStartDate, stepEndDate, index, selectedLanguage })}
	</span>
</div>

<style>
	.row {
		display: flex;
    position: relative;
    gap: 0.5rem;
	}

	.written-plan {
		padding-block: 0.25rem;
	}

	.isInvalid .written-plan{
		color: var(--color-status-error);
	}
	
	.isStepPlaceholder .written-plan{
		color: var(--color-text-muted);
	}

	:where(.isInvalid, .isStepPlaceholder):not(.isLastPlaceholderStep) {
		& .written-plan {
			text-decoration: underline;
			text-decoration-style: wavy;
			text-decoration-skip-ink: none;
			text-decoration-thickness: 1px;
			text-underline-offset: 4px;
		}
	}
</style>
