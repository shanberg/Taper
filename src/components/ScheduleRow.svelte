<script lang="ts">
	import { appStore } from '../stores';
	import FormSegment from './FormSegment.svelte';
	import AddSegmentButton from './AddSegmentButton.svelte';
	import { TaperDate } from '../TaperDate';
	import { getDaysWithDosesForSegment, segmentIsOrAfterPlaceholder, isSegmentInvalid, formatSegmentText, calculateScheduleSummary } from '../utils';
	import List from './List.svelte';
	import ListItem from './ListItem.svelte';

	export let schedule: Schedule;
  export let index: number;
  export let selectedLanguage: Language;
	
	let segmentStartDate: ScheduleDate;
	let segmentEndDate: ScheduleDate;
	
	$: displayMode = schedule.displayMode;
	$: startDate = schedule.startDate;
	$: segments = schedule.segments;
	$: segment = schedule.segments[index];
	$: isSegmentPlaceholder = segment.dose === 0 && segment.daysForDose === 0;
	$: isInvalid = !isSegmentPlaceholder && isSegmentInvalid(segment);
	$: isLastPlaceholderSegment = index === segments.length - 1 && isSegmentPlaceholder;

	$: daysForDose = getDaysWithDosesForSegment(segment, startDate);

	// Calculate the start and end dates
	$: {
		const taperStartDate = new TaperDate(startDate);
		const totalDaysForStartDate =
			segments
				.slice(0, index)
				.reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
			index;
		taperStartDate.incrementByDays(totalDaysForStartDate);
		const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
		taperEndDate.incrementByDays(segment.daysForDose - 1);

		segmentStartDate = taperStartDate.toScheduleDate();
		segmentEndDate = taperEndDate.toScheduleDate();
	}

</script>

<div class="row" class:isInvalid class:isSegmentPlaceholder class:isLastPlaceholderSegment>
  {#if !segmentIsOrAfterPlaceholder(segments, index)}
    <AddSegmentButton
      on:addSegment={() => appStore.insertPlaceholderSegmentBeforeIndex(index)}
    />
  {/if}
  <FormSegment
    {segments}
    {segment}
    {index}
    on:removeSegment={() => appStore.deleteSegmentAtIndex(index)}
    on:change={(event) => appStore.editSegmentAtIndex(index, event.detail)}
  />
  {#if isLastPlaceholderSegment}
    <span class="written-plan summary">
      {calculateScheduleSummary(schedule)}
    </span>
  {:else}
    <span class="written-plan">
      {formatSegmentText({ segment, segmentStartDate, segmentEndDate, index, selectedLanguage })}
    </span>
  {/if}
</div>

{#if displayMode === "doses"}
<div class="spacer">
<List>
{#each daysForDose as day, index}
	<ListItem>Take {day.dose}mg on {new TaperDate(day.date).toYYYYMMDD()}</ListItem>	
{/each}
</List>
</div>
{/if}

<style>
	.row {
		display: flex;
    position: relative;
    gap: 0.5rem;
	}

	.spacer {
		padding-left: 13rem;
		color: var(--color-text-muted);
	}

	.written-plan {
		padding-block: 0.25rem;
	}

	.isInvalid .written-plan{
		color: var(--color-status-error);
	}
	
	.isSegmentPlaceholder .written-plan{
		color: var(--color-text-muted);
	}

	:where(.isInvalid, .isSegmentPlaceholder):not(.isLastPlaceholderSegment) {
		& .written-plan {
			text-decoration: underline;
			text-decoration-style: wavy;
			text-decoration-skip-ink: none;
			text-decoration-thickness: 1px;
			text-underline-offset: 4px;
		}
	}
</style>
