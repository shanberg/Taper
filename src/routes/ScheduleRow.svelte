<script>
  import { createEventDispatcher } from 'svelte';
  import { formatDate } from '../utils';
  export let tableData;
  export let startDate;
  export let row;
  export let index;

  let rowStartDate
  let rowEndDate

  $: rowIsPlaceholder = row.dose === 0 && row.daysForDose === 0
  $: invalid = !rowIsPlaceholder && (row.dose <= 0 || row.daysForDose <= 0);

  // Calculate the start and end dates
  $: {
    rowStartDate = new Date(startDate.getTime());
    const totalDaysForStartDate = tableData.slice(0, index).reduce((acc, curr) => acc + curr.daysForDose, 0) + index;
    rowStartDate.setDate(rowStartDate.getDate() + totalDaysForStartDate);

    rowEndDate = new Date(rowStartDate.getTime());
    rowEndDate.setDate(rowEndDate.getDate() + row.daysForDose);
  }
</script>

{#if !rowIsPlaceholder && !invalid}
  <li>{index === 0 ? "Take" : `Then take`} <b>{row.dose}mg</b> daily for <b>{row.daysForDose} {row.daysForDose === 1 ? 'day' : 'days'}</b> ({formatDate(rowStartDate)}â€“{formatDate(rowEndDate)})</li>
{/if}

<style>
  li { 
  }
  b {
    font-weight: unset;
  }
</style>