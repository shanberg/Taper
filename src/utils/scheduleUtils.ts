import { TaperDate } from '../TaperDate';
import { TEMPLATES, DEFAULT_LANGUAGE_KEY, DEFAULT_TEMPLATE_KEY } from '../consts';
import { isStepInvalid, sumStepsDays, sumStepsDose, isStepPlaceholder } from './stepUtils';
import { formatStepText } from './textUtils';
import { getLanguageFromKey } from './languageUtils';

const PLACEHOLDER_SEGMENT: Step = { dose: 0, daysForDose: 0 };

export function createInitialSchedule(): Schedule {
  return {
    steps: [...TEMPLATES.Default, PLACEHOLDER_SEGMENT],
    startDate: new TaperDate().toScheduleDate(),
    templateKey: DEFAULT_TEMPLATE_KEY,
    languageKey: DEFAULT_LANGUAGE_KEY,
    displayMode: "steps"
  };
}

export function isValidSchedule(schedule: Schedule): boolean {
  if (!schedule) {
    throw new Error("No schedule provided");
  }
  return !schedule.steps.slice(0, schedule.steps.length - 1).some(s => isStepInvalid(s));
}

export function calculateScheduleSummary(schedule: Schedule): string {
  return `${sumStepsDose(schedule.steps)}mg over ${sumStepsDays(schedule.steps)} days`
};

export function calculateStepStartAndEndDates(schedule: Schedule, index: number): StepWithStartEndDate {
  const step = schedule.steps[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.steps
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(step.daysForDose - 1);

  return {
    step,
    stepStartDate: taperStartDate.toScheduleDate(),
    stepEndDate: taperEndDate.toScheduleDate()
  }
}

export function getFormattedListForCopyPaste(schedule: Schedule): string {
  const { steps, languageKey } = schedule;
  const selectedLanguage = getLanguageFromKey(languageKey);

  // Filter out placeholder steps
  const validSteps = steps.filter(step => !isStepPlaceholder(step));

  // Format each step
  const formattedSteps = validSteps.map((step, index) => {
    const { stepStartDate, stepEndDate } = calculateStepStartAndEndDates(schedule, index);

    return formatStepText({
      step,
      stepStartDate,
      stepEndDate,
      index,
      selectedLanguage
    });
  });

  // Combine formatted steps into a single string
  return formattedSteps.join('\n');
}
