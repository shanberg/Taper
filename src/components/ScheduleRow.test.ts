import { describe, test, expect, vi, beforeEach } from 'vitest';
import { render } from '@testing-library/svelte';
import '@testing-library/jest-dom';
import ScheduleRow from './ScheduleRow.svelte'; // Adjust the import path as necessary

// Mock the utility functions and constants
vi.mock('../utils', () => ({
	formatRowText: vi.fn(),
	isRowInvalid: vi.fn()
}));

vi.mock('../consts', () => ({
	LANGUAGES: {
		en: 'English',
		es: 'Spanish',
		zh: 'Mandarin',
		ht: 'Haitian Creole'
	}
}));

import { formatRowText, isRowInvalid } from '../utils';

describe('ScheduleRow', () => {
	let tableData: UITableData;
	let startDate: Date;
	let selectedLanguageKey: string;
	let row: Row;
	let index: number;

	beforeEach(() => {
		tableData = [
			{ dose: 1, daysForDose: 5 },
			{ dose: 2, daysForDose: 3 }
		];
		startDate = new Date('2024-06-01');
		selectedLanguageKey = 'en';
		row = { dose: 1, daysForDose: 5 };
		index = 0;

		(formatRowText as jest.Mock).mockImplementation(
			({ row, rowStartDate, rowEndDate, index, selectedLanguageKey }) => {
				return `Row ${index + 1}: ${row.dose} dose(s) from ${rowStartDate.toDateString()} to ${rowEndDate.toDateString()}`;
			}
		);

		(isRowInvalid as jest.Mock).mockImplementation((row) => row.dose < 0 || row.daysForDose < 0);
	});

	test('renders correctly', () => {
		const { container } = render(ScheduleRow, {
			tableData,
			startDate,
			selectedLanguageKey,
			row,
			index
		});
		expect(container).toMatchSnapshot();
	});

	test('applies isInvalid class when the row is invalid', () => {
		row = { dose: -1, daysForDose: 5 };
		const { container } = render(ScheduleRow, {
			tableData,
			startDate,
			selectedLanguageKey,
			row,
			index
		});
		expect(container.querySelector('li')).toHaveClass('isInvalid');
	});

	test('applies isRowPlaceholder class when the row is a placeholder', () => {
		row = { dose: 0, daysForDose: 0 };
		const { container } = render(ScheduleRow, {
			tableData,
			startDate,
			selectedLanguageKey,
			row,
			index
		});
		expect(container.querySelector('li')).toHaveClass('isRowPlaceholder');
	});

	test('does not render the last placeholder row', () => {
		tableData = [
			{ dose: 1, daysForDose: 5 },
			{ dose: 0, daysForDose: 0 }
		];
		row = { dose: 0, daysForDose: 0 };
		index = 1;
		const { container } = render(ScheduleRow, {
			tableData,
			startDate,
			selectedLanguageKey,
			row,
			index
		});
		expect(container.querySelector('li')).not.toBeInTheDocument();
	});

	test('formats the row text correctly', () => {
		const { getByText } = render(ScheduleRow, {
			tableData,
			startDate,
			selectedLanguageKey,
			row,
			index
		});
		expect(getByText(/Row 1: 1 dose\(s\) from/)).toBeInTheDocument();
	});
});
