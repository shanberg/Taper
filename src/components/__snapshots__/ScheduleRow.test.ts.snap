// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScheduleStep > renders correctly 1`] = `
<div>
  <li
    class="svelte-1hlpwdm"
  >
    Step 1: 1 dose(s) from Sat Jun 01 2024 to Wed Jun 05 2024
  </li>
  
</div>
`;
